Mangle synthesized error structs as a variant of the imported enum

jrose-apple opened the pull request:
When importing a C enum with the ns_error_domain attribute, we synthesize a struct containing an NSError object to represent errors in that domain. That synthesized struct should have a mangled name that ties it to the original C enum, if we want it to be stable, and now it does.

Before: $SSC7MyErrorV (a normal struct, which is a lie)
After: $SSC11MyErrorCodeLeV
```
kind=Global
  kind=Structure
    kind=Module, text="__C_Synthesized"
    kind=RelatedEntityDeclName, text="e"
      kind=Identifier, text="MyErrorCode"
```

More rdar://problem/24688918.

jrose-apple committed:
Mangle synthesized error structs as a variant of the imported enum

When importing a C enum with the ns_error_domain attribute, we
synthesize a struct containing an NSError object to represent errors
in that domain. That synthesized struct should have a mangled name
that ties it to the original C enum, if we want it to be stable, and
now it does.

Before: $SSC7MyErrorV (a normal struct, which is a lie)
After: $SSC11MyErrorCode13ns_error_enumLLV
  kind=Global
    kind=Structure
      kind=Module, text="__C_Synthesized"
      kind=PrivateDeclName
        kind=Identifier, text="ns_error_enum"
        kind=Identifier, text="MyErrorCode"

Using the "private discriminator" feature allows us to pack in extra
information about the declaration without changing the mangling
grammar, and without stepping on anything the importer is using.

More rdar://problem/24688918

jrose-apple commented:
@swift-ci Please test

jrose-apple commented on lib/ClangImporter/ImportDecl.cpp:
```diff
@@ -2613,6 +2613,17 @@ namespace {
+            nameForMangling = decl->getName();
+            discriminatorForMangling = ERROR_ENUM_MANGLING_KEY;
+          }
+          errorWrapper->getAttrs().add(new (C) CustomTypeNameManglingAttr(C.getIdentifier(nameForMangling), C.getIdentifier(discriminatorForMangling)));
```
Oops, that definitely breaks the 80-column limit. Will fix tomorrow.

rjmccall commented:
We can't just use the C-namespace name without other decoration?  Does that break too much reflection and so on?

jrose-apple commented:
We use the C namespace normally for the error code, something like "$SSo7MyErrorV" (I made the executive decision to mangle all C tag types as "V" rather than to preserve a "V"/"O" distinction.) This is about the struct that we conjure up out of thin air to wrap the NSError that goes with the code. I thought about some other options:

- Just change the module: "$S**SC**7MyErrorV". Subtle but sufficient at the moment.
- Change the node kind: "$SSC7MyError**O**" (or maybe use this one for the code and the original one for the struct).
- Invent a different name based on the original name and teach the importer to find it: "$SSC16MyError__wrapperV"

Reusing the private discriminator was the one I liked the best, though. It's a solution that scales to anything else we might want to synthesize in the future, and it handles the anonymous enum problem that I didn't even think of until I started implementing this (that is, the difference between the ordinary and tag namespaces). And we know we're not using it for anything for imported types because those never have anything like file-private manglings. (Well, not in C, anyway. It may or may not make sense to use something like this for types defined in anonymous namespaces in C++ header files.)

I like this solution, but I'm not too attached to it if you have another idea.

rjmccall commented:
Thank you for the explanation.  My concern is purely rooted in the way that this leaves the synthesized struct name in the mangling, which both inflates manglings (unfortunately especially because we expect people to work with the struct more often) and potentially affects stability.  Is there some way you can invert this, using the C enum name as the primary name and then doing something discriminator-ish to indicate that this is a synthesized error type for the domain?

jrose-apple commented:
Oh, this isn't using the synthesized struct name! It is using the C enum's name (note "MyError" vs. "MyErrorCode" in the original example, which I forgot to carry over into my explanation comment). But it's definitely a painfully-larger mangled name. We could at least shrink the discriminator to a single letter if you think that's better.

rjmccall commented:
Oh, I see, ns_error_enum is literal in this.  Sorry for the misunderstanding!

Yes, please use a shorter discriminator.  It looks like ``L`` creates a namespace of name-modifiers, of which only ``LL``, ``Ll``, and ``L[0-9]+`` are claimed, so you could just do something like ``LE``.

jrose-apple commented:
Oh, you mean to actually change the mangling grammar? I was trying to avoid that for simplicity's sake; I was thinking "1eLL" and "1ELL" for the two variants here. That way we won't need to change the demangler anymore if we add another one.

rjmccall commented:
Is not changing the demangler important to this task for some reason?

jrose-apple commented:
I just don't think this will be the last importer-synthesized type, and therefore I wanted to leave space for the future. That is, we could change the demangler *now* (and I did have to change type reconstruction, for example), but I don't want to change it again next year.

eeckstein commented:
You can use 'L' with a character suffix other than [lL0-9_], like 'Le' and 'LE'.
This leaves plenty of space for future extensions and IMO is cleaner than to encode a flag as an pseudo identifier.

jrose-apple commented:
That doesn't keep us from having to change the demangler if we add a new flag, since the various 'L' manglings don't all have the same kind of payload. ('LE' would have none besides the name, for example.)

I really think we're going to get to a point where the demangler or something else that depends on the node tree lives in Apple's OS, and therefore we won't be able to change the mangling structure for anything that needs to backwards-deploy. That's why I want something extensible.

eeckstein commented:
Not necessarily. The demangler could accept any character after 'L'. The question how should that show up in the demangled name?

jrose-apple commented:
Accepting any character after L isn't good enough because LL, Ll, and L0 all have different structure. Unless we want to claim that all remaining L* don't have structure.

jrose-apple commented:
Erik came to talk to me in person and suggested claiming a *range* of discriminators without claiming *all* of the remaining letters. I'll see what that looks like.

jrose-apple commented:
I ended up not being going too far down that route because I'd have to add it to the old mangling too. Here's an alternate version that uses `LE` as a (mandatory, canonical) abbreviation for `2$ELL`. @rjmccall, what do you think?

jrose-apple commented:
@swift-ci Please smoke test

rjmccall commented:
I don't understand the purpose of formalizing it as an abbreviation.  It's canonical, i.e. it's illegal to use the other form, so why even think of it as having any relation to the other form?  Just say that these specific new forms are reserved for the purpose of identifying types that are in various ways related to the original type, add demangler support for them so that it at least doesn't fail, and call it a day.

I don't see any way that demangler-derived but OS-deployed tools like Doug's mangling-to-metadata function can possibly do the right thing on these types without understanding the derivation anyway.

jrose-apple commented:
It's related to the other form because otherwise I'd have to invent a new form in the old mangling too. Which I certainly could, if you think that's better, but it seems weird to keep messing with the old mangling.

(On the other hand, that's going to be a problem forever, isn't it? We'd have to decide that generic classes don't need to use the old mangling at all, otherwise, or make it conditional on what the generic arguments are. Darn NSKeyedArchiver backwards-compatibility fears.)

rjmccall commented:
Why do you need it in the old mangling?  I think that only exists for specific old formations now.  Most of the new manglings we've added aren't getting added there.

(But if you need to add it, generally the old mangling version of the new mangling is just a prefix-encoding of what the new mangling does as a suffix.)

rjmccall commented:
I guess the polite thing to do is to post a follow-up rather than editing the old response into oblivion.

Yes, we should absolutely make sure that the keyed-archiver stuff is only used for the basic type forms that were allowed by the old mangling; I do not think we want to indefinitely support the entire old mangling, or to support an odd mish-mash of both worlds there.  I kindof expected that we'd done that already, frankly.

jrose-apple commented:
Okay, I'll look into that tomorrow.

jrose-apple commented:
Okay, here's a version that uses a new custom node kind instead of piggybacking on private discriminators (except when encoding to the old mangling). What do you two think?

@swift-ci Please test

jrose-apple commented:
Note that I didn't end up doing anything to curtail the use of the old mangling; @eeckstein said it was fine to just do a "very-unlikely-to-have-conflicts" map. Whether we can actually stop using it ought to be a separate PR anyway.

rjmccall commented on docs/ABI/Mangling.rst:
```diff
@@ -216,19 +216,28 @@ Entities
+  decl-name ::= identifier 'L' RELATED-DISCRIMINATOR  // related declaration
+
+  RELATED-DISCRIMINATOR ::= [a-j]
+  RELATED-DISCRIMINATOR ::= [A-J]
```
Please document the fact that you've claimed some of these here.

jrose-apple replied:
I don't consider them "claimed". Their meaning is context-dependent. "The Clang importer is using 'LE' and 'Le' for top-level values under the 'SC' module" doesn't seem like part of the mangling schema or general ABI to me.

rjmccall commented on docs/ABI/Mangling.rst:
```diff
@@ -216,19 +216,28 @@ Entities
+Twenty operators of the form 'LA', 'LB', etc. are reserved to described
+entities related to the entity whose name is provided. For example, 'LE' and
+'Le' in the "SC" module are used to represent the structs synthesized by the
+Clang importer for various "error code" enums.
```
Why both?

jrose-apple replied:
Enums in the tag namespace vs. anonymous enums with a typedef-for-linkage, to catch the case where someone obnoxiously defines the same enum in both namespaces. But that's not part of the ABI.

rjmccall commented on include/swift/AST/Attr.h:
```diff
@@ -1199,6 +1199,29 @@ class RestatedObjCConformanceAttr : public DeclAttribute {
+    assert(!relatedEntityKind ||
+           (relatedEntityKind >= 'a' && relatedEntityKind <= 'j') ||
+           (relatedEntityKind >= 'A' && relatedEntityKind <= 'J') &&
+           "only [A-Ja-j] are supported as related entity kinds");
```
It seems wrong to actually use the mangling in the representation.  Maybe have a Kind and then map to/from the char?

jrose-apple replied:
I thought about that, but then you just end up with an enum of 20 cases with no intrinsic meaning (because "context-dependent"). I could have restricted it to a number 1-20 or 0-19, but that's less debuggable.

rjmccall replied:
Why would it need to be context-dependent?  You can use a semantically meaningful enum and then map it to the context-dependent case.

jrose-apple replied:
John and I talked about this in person and he convinced me to make something more semantic here, like ClangImporterSynthesizedType.

rjmccall commented on lib/AST/ASTMangler.cpp:
```diff
@@ -635,6 +645,12 @@ void ASTMangler::appendDeclName(const ValueDecl *decl) {
+  if (customNameAttr && customNameAttr->RelatedEntityKind) {
+    return appendOperatorParam("L",
+                               StringRef(&customNameAttr->RelatedEntityKind,
+                                         1));
```
Seems to me that, even in your representation, there should be an accessor on the attribute that returns a StringRef.

jrose-apple replied:
Fair enough.

rjmccall commented on lib/ClangImporter/ImportEnumInfo.h:
```diff
@@ -31,6 +31,9 @@ class MacroInfo;
 namespace importer {
 
+static const char ERROR_ENUM_MANGLING_KEY = 'e';
+static const char ERROR_ENUM_ANON_MANGLING_KEY = 'E';
```
Comments?

eeckstein reviewed (approved):
mangling changes lgtm!

jrose-apple committed:
[Mangling] Define "related entity" operators 'LA'...'LJ'

(and 'La'...'Lj')

Use this for the synthesized structs for error enums, as described in
the previous commit, instead of reusing the "private discriminator"
feature. I left some space in the APIs for "related entity kinds" that
are longer than a single character, but I don't actually expect to use
it any time soon. It's mostly just easier to deal with StringRef than
with a bare char.

Note that this doesn't perfectly round-trip to the old mangling; I had
it treat these nodes as private discriminators with a prefixed "$"
instead. We don't depend on that for anything, though.

jrose-apple commented:
Addressed some of John's feedback, but pushed back against other parts. Waiting to see if he's convinced.

@swift-ci Please test

jrose-apple committed:
Use a semantic ClangImporterSynthesizedTypeAttr for error structs

...rather than the ad hoc CustomTypeNameManglingAttr I was using
before. As John pointed out, the AST should be semantic wherever
possible.

We may someday want to get out of this being an attribute altogether,
or duplicating information that's available in the original Clang
node, by actually storing a reference to that node somewhere. This is
tricky and mixed up with deciding what hasClangNode() or
getClangDecl() would mean, though, so for now the attribute just
carries the information we need.

jrose-apple commented:
@rjmccall, any further comments?

@swift-ci Please test

rjmccall commented:
No, thank you, the design looks great.
